spec/e2e/association_proxy_spec.rb

```diff
@@ -107,6 +115,26 @@
     end
   end
 
+  it 'Queries limited times in depth two loops with deep with_associations' do
+    Student.create.lessons << science
+    Student.create.lessons << science
+    expect_queries(1) do
+      science.students.with_associations(lessons: :exams_given).each do |student|
+        student.lessons.each do |lesson|
+          lesson.exams_given.to_a
+        end
+      end
+    end
+  end
+
+  it 'Queries only one time when there are some empty associations' do
+    Student.create.lessons << science
+    Student.create.lessons += [science, Lesson.create]
+    expect_queries(1) do
+      science.students.with_associations(lessons: :exams_given).flat_map(&:lessons).flat_map(&:exams_given)
+    end
+  end
```

Comment:
I see that you are getting the relationships in the query as well.  Would these be available by doing an `each_rel`?  If so, I think there's at least one spec missing here.

I also wonder if it makes sense to have another function which doesn't return the relationship for the sake of not loading as much data.  Probably that's a project for another day...